search/searcher/search_nested.go

```diff
@@ -0,0 +1,154 @@
+//  Copyright (c) 2019 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package searcher
+
+import (
+	"github.com/blevesearch/bleve/index"
+	"github.com/blevesearch/bleve/search"
+)
+
+type depthFilterContext struct {
+	termAPMap  map[string][]search.ArrayPositions
+	apLists    [][]search.ArrayPositions
+	arrayDepth int
+}
+
+func NewNestedArraySearcher(indexReader index.IndexReader, depth int,
+	qsearchers []search.Searcher, options search.SearcherOptions) (
+	search.Searcher, error) {
+	conjunctionSearcher, err := NewConjunctionSearcher(indexReader, qsearchers,
+		options)
+	if err != nil {
+		return nil, err
+	}
+
+	rv := depthFilterContext{arrayDepth: depth}
+
+	return NewFilteringSearcher(conjunctionSearcher,
+		rv.buildArrayDepthFilter()), nil
+}
+
+// buildArrayDepthFilter returns a filter that checks whether
+// all the search terms are sharing any common array position in
+// the document match.
+func (c *depthFilterContext) buildArrayDepthFilter() FilterFunc {
+	return c.checkArrayDepth
+}
+
+func (c *depthFilterContext) checkArrayDepth(d *search.DocumentMatch) bool {
+	if d == nil || len(d.FieldTermLocations) == 0 {
+		return false
+	}
+
+	// reset and resuse
+	if len(c.termAPMap) == 0 {
+		c.termAPMap = make(map[string][]search.ArrayPositions, len(d.FieldTermLocations))
+	} else {
+		for k := range c.termAPMap {
+			list := c.termAPMap[k]
+			c.termAPMap[k] = list[:0]
+		}
+	}
+
+	termList := make(map[string]struct{}, 1)
+	for _, ftl := range d.FieldTermLocations {
+		if len(ftl.Location.ArrayPositions) == c.arrayDepth {
```

Comment:
What is special about the FTL having the exact depth requested as the query array depth?  I thought this was designed such that you could specify the minimum depth that you care about.  In the book example, depth 1 might mean same chapter, depth 2 might mean same page, etc.  I could query for depth 1, meaning I want to find matches that are in the same chapter, but the hits all actually have depth 3, because the matching terms occur inside pages, which are nested inside of chapters...

Reply:
It was designed for exact depth. Now in book sample documents under tests, 
{book [chapters : [pages]]} . results in array Positions slices of depth 2, not depth 3.